src/Interface/Wx/MapViewer.pm

```diff
@@ -310,30 +311,33 @@ sub setMapDir {
 sub parsePortals {
 	my $self = shift;
 	my $file = shift;
-	return unless (-r $file);
-	open FILE, "< $file";
-	$self->{portals} = {};
-	while (my $line = <FILE>) {
-		next if $line =~ /^#/;
-		$line =~ s/\cM|\cJ//g;
-		$line =~ s/\s+/ /g;
-		$line =~ s/^\s+|\s+$//g;
-		my @args = split /\s/, $line, 8;
-		if (@args > 5) {
-			$self->{portals}->{$args[0]} = [] unless defined $self->{portals}->{$args[0]};
-			push (@{$self->{portals}->{$args[0]}}, {
-				x => $args[1],
-				y => $args[2],
-				destination => {
-					field => $args[3],
-					x => $args[4],
-					y => $args[5],
-				},
-				zeny => $args[7],
-			});
+	delete $self->{portals};
+
+	foreach my $portal (keys %portals_lut) {
+		foreach my $dest (keys %{$portals_lut{$portal}{dest}}) {
+			next if $portals_lut{$portal}{dest}{$dest}{map} eq '';
+			if ($portals_lut{$portal}{dest}{$dest}{steps}) {
+				# this is a Warp NPC
+				push (@{$self->{portals}->{$portals_lut{$portal}{source}{map}}}, {
+					x => $portals_lut{$portal}{source}{x},
+					y => $portals_lut{$portal}{source}{y},
+					npcType => '1',
+				});
+			last;
+			} else {
+				# this is a portal
+				push (@{$self->{portals}->{$portals_lut{$portal}{source}{map}}}, {
+					x => $portals_lut{$portal}{source}{x},
+					y => $portals_lut{$portal}{source}{y},
+					destination => {
+						field => $portals_lut{$portal}{dest}{$dest}{map},
+						x => $portals_lut{$portal}{dest}{$dest}{x},
+						y => $portals_lut{$portal}{dest}{$dest}{y},
+					},
+				});
+			}
 		}
 	}
-	close FILE;
```

Comment:
will openkore keep the file open?

Reply:
the is no file being opened anymore

Reply:
yes, now we do not reopen the file. We use a ready-made global hash `%portals_lut`